proposals/heading-component.md

```diff
@@ -0,0 +1,73 @@
+# baseui/heading
+
+This is a proposal to create a new component `baseui/heading` that can serve as a smarter alternative to `baseui/typography/h*` family components. If this proposal gets approved and implemented we should slowly deprecate `baseui/typography/h*` components.
+
+## Motivation
+
+Headings are absolutely critical for screen reader users and it's important to correctly set their levels. For example, you should never skip a level (having h2 and not h1 on the same page).
+
+The problem is that levels are a question of context. When you move components with headings around, you need to check if levels still make sense and adjust them if not. That's a lot of mundane work and developers often rather take shortcuts such as using `h1` element only.
+
+There is proposal for an `<h>` element and [document outline algorithm](http://adrianroselli.com/2016/08/there-is-no-document-outline-algorithm.html) to solve this problem once for all but there is no browser support yet. Fortunately, we [can use React Context](https://medium.com/@Heydon/managing-heading-levels-in-design-systems-18be9a746fa3) to implement our own document outline algorithm.
+
+## Usage
+
+```jsx
+import * as React from 'react';
+import {Heading} from 'baseui/heading';
+export default () => (
+  <Heading heading="Base Web">
```

Comment:
I find the api for the `Heading` component confusing. It's hard to imagine what the resulting markup looks like without viewing an example like what's provided below. While not knowing the markup is common when working with react components, I'm concerned when it has a very similar mapping to html `<Hn>` elements.

I would prefer to see a similar API to http://adrianroselli.com/2016/08/there-is-no-document-outline-algorithm.html where there are two components. One for `Headings` and another for `Sections`. This way, each component is semantically consistent with its related html element

Reply:
Yea, I've actually almost submitted this proposal with `Section` and `Heading` components since that's what [this article](https://medium.com/@Heydon/managing-heading-levels-in-design-systems-18be9a746fa3) also proposes. 

However, I was not convinced it's not even more confusing because the `Section` component doesn't render the `<section>` element (and it probably should not). It **only manages the context value/leveling** and then it becomes somewhat unnecessary because you can do the same thing with a single component.

But I agree that it might be also surprising to see `<Heading heading="foo">Related text</Heading>`.

Also should you be able to do both:

```jsx
<Heading>
<Section>
  <Heading>
</Section>
```

and

```jsx
<Section>
  <Heading>
  <Section>
    <Heading>
  </Section>
</Section>
```

If so, should the ouput be the same?

Reply:
Good point about not rendering a `section` element. Maybe an alternative name is more appropriate? `HeadingLevel`? 

My thought is that `Section` would be a context provider. Could `Heading` throw if no context is found? That way, only the second option would be technically possible.